File: 692-top-k-frequent-words/top-k-frequent-words.cpp
class Solution {
public:
    
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> mp;
        for(string s: words)
            mp[s] += 1;
        vector<string> ans;
        auto cmp = [](const pair<int, string> &a, const pair<int, string> &b){
            if(a.first == b.first)
                return a.second > b.second;
            return a.first < b.first;
        };
        priority_queue<pair<int, string>, vector<pair<int, string>>, decltype(cmp)> pq(cmp);
        for(auto t : mp){
            pq.push({t.second, t.first});
        }
        while(k--){
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
    }
};

File: 692-top-k-frequent-words/top-k-frequent-words.java
class Solution {
    class Pair {
        Integer first;
        String second;

        Pair(Integer first, String second){
            this.first = first;
            this.second = second;
        }
    }
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> mp = new HashMap<>();
        int n = words.length;
        for(int i = 0; i < n; i++){
            mp.put(words[i], mp.getOrDefault(words[i], 0) + 1);
        }
       PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> {
            if (a.first != b.first) {
                return a.first.compareTo(b.first);
            } else {
                return b.second.compareTo(a.second); // reverse lex order
            }
        });
        for(Map.Entry<String, Integer> item : mp.entrySet()){
            pq.add(new Pair(item.getValue(), item.getKey()));
            if(pq.size() > k)
                pq.poll();
        }
        
        
        List<String> ans = new ArrayList<>();
        while(!pq.isEmpty()){
            ans.add(pq.poll().second);
        }
        Collections.reverse(ans);
        return ans;
    }
}